packages/editor/src/editor.ts

```diff
@@ -0,0 +1,187 @@
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import { once } from 'events';
+import { Readable } from 'stream';
+import spawn from 'cross-spawn';
+import { v4 as uuidv4 } from 'uuid';
+
+import type { ShellUserConfig, MongoshBus } from '@mongosh/types';
+import { signatures, ShellPlugin, ShellInternalState, TypeSignature } from '@mongosh/shell-api';
+
+export interface EditorOptions {
+  input: Readable;
+  vscodeDir: string,
+  tmpDir: string,
+  internalState: ShellInternalState;
+  makeMultilineJSIntoSingleLine: any;
+}
+
+export class Editor implements ShellPlugin {
+  _input: Readable;
+  _vscodeDir: string;
+  _tmpDir: string;
+  _tmpDocName: string;
+  _tmpDoc: string;
+  _internalState: ShellInternalState;
+  _makeMultilineJSIntoSingleLine: any;
+  load: (filename: string) => Promise<void>;
+  require: any;
+  config: { get<T extends keyof ShellUserConfig>(key: T): Promise<ShellUserConfig[T]> };
+  print: (...args: any[]) => Promise<void>;
+  npmArgv: string[];
+
+  constructor({ input, vscodeDir, tmpDir, internalState, makeMultilineJSIntoSingleLine }: EditorOptions) {
+    const { load, config, print, require } = internalState.context;
+    this._input = input;
+    this._vscodeDir = vscodeDir;
+    this._tmpDir = tmpDir;
+    this._tmpDocName = `edit-${uuidv4()}`;
+    this._tmpDoc = '';
+    this._internalState = internalState;
+    this._makeMultilineJSIntoSingleLine = makeMultilineJSIntoSingleLine;
+    this.load = load;
+    this.config = config;
+    this.print = print;
+    this.require = require;
+    this.npmArgv = [];
+
+    // Add edit command support to shell api.
+    const wrapperFn = (...args: string[]) => {
+      return Object.assign(this.runEditorCommand(args), {
+        [Symbol.for('@@mongosh.syntheticPromise')]: true
+      });
+    };
+    wrapperFn.isDirectShellCommand = true;
+    wrapperFn.returnsPromise = true;
+    (internalState.shellApi as any).edit = internalState.context.edit = wrapperFn;
+    (signatures.ShellApi.attributes as any).edit = {
+      type: 'function',
+      returnsPromise: true,
+      isDirectShellCommand: true
+    } as TypeSignature;
+  }
+
+  activate(): void {
+    this._internalState.registerPlugin(this);
+  }
+
+  // In the case of using VSCode as an external editor,
+  // we should detect whether the MongoDB extension is installed and open a .mongodb file.
+  // In all other cases we should open a .js file.
+  async getExtension(cmd: string): Promise<string> {
+    if (!cmd.includes('code')) {
```

Comment:
We will need a stricter check here, because this would also match things like `xcode` or `codeblocks` :)